llvm/lib/Target/EraVM/EraVMPostCodegenPrepare.cpp

```diff
@@ -326,11 +317,213 @@ bool EraVMPostCodegenPrepare::rearrangeOverflowHandlingBranches(Function &F) {
+
+    // Add instruction and update the last split index if needed.
+    void addInst(Instruction *I) {
+      assert(isa<ConstantInt>(I->getOperand(1)) && "Expected constant operand");
+
+      if (LastSplitIdx == UINT64_MAX ||
+          !I->comesBefore(Insts[LastSplitIdx].first))
+        LastSplitIdx = Insts.size();
+
+      Insts.push_back({I, cast<ConstantInt>(I->getOperand(1))->getValue()});
+    }
+
+    // Get the last instruction that will be used to split the live range.
+    InstInfo &getLastSplitInst() {
+      assert(LastSplitIdx != UINT64_MAX && "Expected valid index");
+      return Insts[LastSplitIdx];
+    }
+  };
+
+  bool Changed = false;
+  for (auto *L : LI) {
+    for (auto &PN : L->getHeader()->phis()) {
+      if (PN.getNumUses() <= NumOfPHIUsesToSplitLiveRanges)
+        continue;
+
+      DomTreeNode *RootNode = DT.getNode(L->getHeader());
+      if (!RootNode)
+        continue;
+
+      MapVector<BasicBlock *, InstSplitInfo> Splits;
+      for (auto *U : PN.users()) {
+        // Only collect add instructions with constant operands in this loop.
+        auto *UI = cast<Instruction>(U);
+        if (!UI || UI->getOpcode() != Instruction::Add ||
+            !isa<ConstantInt>(UI->getOperand(1)) ||
+            LI.getLoopFor(UI->getParent()) != L)
+          continue;
+
+        auto [Itr, _] = Splits.insert({UI->getParent(), {}});
+        Itr->second.addInst(UI);
+      }
+
+      // Skip if there are no at least two blocks where we can split
+      // the live range.
+      if (Splits.size() < 2)
+        continue;
+
+      // Split ranges across blocks. This is done by finding the nearest
+      // dominator block, from which we can split the live range.
+      for (auto &[BB, Infos] : Splits) {
+        DomTreeNode *Node = DT.getNode(BB);
+        if (!Node)
+          continue;
+
+        // Find the nearest dominator, from which we can split the live range.
+        InstSplitInfo::InstInfo *NearestDominator = nullptr;
+        while ((Node = Node->getIDom())) {
+          auto *I = Splits.find(Node->getBlock());
+          if (I != Splits.end()) {
+            // We found the nearest dominator block, so take the last
+            // instruction from it.
+            NearestDominator = &I->second.getLastSplitInst();
+            break;
+          }
+
+          // Bail out if we reached the start of the loop.
+          if (Node == RootNode)
+            break;
+        }
+
+        // If we didn't find any dominator, skip this BB.
+        if (!NearestDominator)
+          continue;
+
+        // Update operands in this block with the nearest dominator.
+        // Basically, we are doing the following transformation:
+        //   %Dominator = add %In, DomImm
+        //   %I         = add %In, Imm
+        // to
+        //   %Dominator = add %In, DomImm
+        //   %I         = add %Dominator, Imm - DomImm
+        auto [DomInst, DomImm] = *NearestDominator;
+        for (auto &Info : Infos.Insts) {
```

Comment:
Big mental complexity of the function (nested loops of depth 4). Perhaps it is worth extracting the rewrite logic and other parts into separate functions and calling them from the loop.

Reply:
Split this code into 3 functions.